### ISSUE TITLE
[12.0] .../xmlrpc/db: 404 NOT FOUND

### ISSUE THREAD

**Issue by gaikaz** [Yenthe666/auto_backup#122]
Full error in the popup:
```
Error ! No such database exists!

Error details:
<ProtocolError for localhost:8069/xmlrpc/db: 404 NOT FOUND>
```

How to reproduce:
- Odoo version 12.0
- Multiple DBs on one Odoo instance (/web/database/selector must give multiple options)
- Back-ups configuration:
*Host:* **localhost** (default)
*Database:* **[Current DB]** (default)
*Port:* **8069** (default)

The problem is caused by the multiple databases. At first I was very confused as to why my browser seems to be able to open the `/xmlrpc/db` uri and server gives a xml response (even if it is a faulty one), but trying to use `curl` ends up in an actual 404. Then I figured out that it remembers the `session_id` from normal use of Odoo and so sends that to the server as well. Removing the cookie results in server giving 404 to the browser as well.

To check my diagnosis further, I tried running Odoo with `dbfilter` set to only find one of the databases and indeed this helps with opening `/xmlrpc/db` from anywhere, as it can automatically figure out the database and assign a new session. 

It seems that a lot has changed how Odoo handles RPC requests in V12. From what I've gathered looking at V9-12 code, it looks like in V12 (maybe V11 too) Odoo starts handling RPC (HTTP) requests more like any other module would - with proper controllers and `@route` directives in them.
So my guess is that this includes session checking somewhere deeper in the code. Maybe this is an actual issue to be reported to Odoo repositories, as the `/xmlrpc/db` -> `list` method kind of becomes redundant.

Hopefully I provided enough information about the issue to save as much of your time as possible. :)

**Comment by Yenthe666** [comment]
Hi @gaikaz,

Thank you so much for the very detailed report, that's to awesome! I'll need some extra time to look into this as it it quite specific and big to test. I'll try to get back to you soon to see if I can figure this out. ;)

Remind me in two weeks if you didn't get a response please. I sometimes tend to miss or forget some, sorry if that happens. ;)

**Comment by gaikaz** [comment]
@Yenthe666 , kind of feel bad for doing this, but those 2 weeks passed :) Could you look into this when you have time?

**Comment by Yenthe666** [comment]
Sorry, I didn't get around to it, I've been sick a few days too.
I've pinned it into my list of "Saved for later" on Github - I've got 20 outstanding items there - that should remind me to check into this one. 
Just searching for some extra time 🙏

**Comment by gr22** [comment]
I can confirm the problem in V12.0 CE

**Comment by gaikaz** [comment]
@Yenthe666 , what's your schedule these days? :) Could you estimate when this could be looked into?
I would help as much as I can.

**Comment by Yenthe666** [comment]
@gaikaz can you perhaps remind me half next week? I'm still as busy but I'll try to squeeze in a check next week. Very sorry.

**Comment by gaikaz** [comment]
@Yenthe666, auto_backup issue reminder 🙊
I really hate doing this 😄

**Comment by Yenthe666** [comment]
@gaikaz no problem, I finally got around to testing it.
I created two databases, both scheduled them to backup and logged out everywhere from all Odoo databases.
Result in the logs from db 1:
```
2019-04-03 07:40:15,160 6424 INFO demo12subscriptions odoo.addons.base.models.ir_cron: Starting job Backup scheduler. 
2019-04-03 07:40:15,190 6424 INFO None werkzeug: 127.0.0.1 - - [03/Apr/2019 07:40:15] "POST /xmlrpc/db HTTP/1.1" 200 - 2 0.005 0.012
2019-04-03 07:40:15,193 6424 INFO demo12subscriptions odoo.service.db: DUMP DB: demo12subscriptions format zip 
```

Result in the logs from db2:
```
2019-04-03 07:41:17,674 6424 INFO demoworkshop odoo.addons.base.models.ir_cron: Starting job Backup scheduler. 
2019-04-03 07:41:17,700 6424 INFO None werkzeug: 127.0.0.1 - - [03/Apr/2019 07:41:17] "POST /xmlrpc/db HTTP/1.1" 200 - 2 0.005 0.010
2019-04-03 07:41:17,703 6424 INFO demoworkshop odoo.service.db: DUMP DB: demoworkshop format zip 
```

I also get two backups on my disk.
Do you have further configurations or anything? I'm not able to reproduce it.

**Comment by gaikaz** [comment]
@Yenthe666 I'll try my case again later today on a fully fresh Odoo install.
Maybe this really was core issue that was fixed.

**Comment by susankc** [comment]
When i try to save a file with appropriate information i came with error below but this error does not exist in localhost. Can you tell me why this is happening? 

Error ! No such database exists!

Error details:
<ProtocolError for localhost:8016/xmlrpc/db: 404 NOT FOUND>

**Comment by Yenthe666** [comment]
Your database/Odoo probably runs on another port @susankc 😉

**Comment by gaikaz** [comment]
Good news everybody! We have a delivery to... I mean, I figured it out! 😃

The problem was incorrect use of `--load=...` launch argument.
Because I use `letsencrypt` module, it has to be loaded server-wide, which is done with this argument.

My setup used: `--load=web,letsencrypt`

Module `web` here is specified, because it is a default of this argument and we still want to load it.
On Odoo 10.0 the default was `web,web_kanban`.
On Odoo 11.0 it was changed to `web`.
On Odoo 12.0 it is now `base,web`

As we can see, Odoo 12.0 now has `base` loaded this way. It means, that module `base` used to be loaded server-wide (on all databases as a sort of a "singleton") automatically. Therefore database independent HTTP requests like `/xmlrpc/db` used to work no matter if database was selected.
On my setup `base` was loaded on each database separately and that's why I wouldn't get a response from the server on `/xmlrpc/db` without first selecting a database. @gr22, tagging you here, as you probably have something similar done with your server 😉

**TL;DR:**
If you are using `--load=...` launch argument with Odoo 12.0, you have to make sure you have `base` module loaded with it.

**Comment by Yenthe666** [comment]
Aha! 🎉 Hey that was quite an interesting and insightful comment to read. It are these kind of things that make sense but only after you've solved them 😆 
I'm happy you've figured it out!

**Comment by susankc** [comment]
No my database is in same port i cannot even save my file. I came upto same error again and again. @Yenthe666

**Comment by Yenthe666** [comment]
Please add more info such as a screenshot of the configuration.
Are you sure you don't have a `db-filter` set or `ngnix` configurations?

**Comment by susankc** [comment]
My configuration
[options]
addons_path = 
db_host = False
db_name = False
db_port = False
db_user = odoo12
db_password = False
dbfilter = ^%h$
geoip_database = /usr/share/GeoIP/GeoLiteCity.dat
limit_memory_hard = 1677721600
limit_memory_soft = 629145600
limit_request = 8192
limit_time_cpu = 800
limit_time_real = 1600
list_db = True
log_db = False
log_db_level = warning
log_handler = :INFO
xmlrpc = True
xmlrpc_interface = 127.0.0.1
xmlrpc_port = 8016
 and my nginx setup
#odoo server
upstream odoo {
        server 127.0.0.1:8016;
}
upstream odoo {
        server 127.0.0.1:8078;
}

**Comment by susankc** [comment]
I set dbfilter to  ^%h$ @Yenthe666

**Comment by Yenthe666** [comment]
Why two upstreams on different ports?
upstream odoo {
server 127.0.0.1:8016;
}
upstream odoo {
server 127.0.0.1:8078;
}

The rest looks fine honestly. Perhaps the configuration in Odoo itself for the backup is wrong too?

**Comment by susankc** [comment]
One port for long polling and one for xmlrpc  actually it was 
#odoo server
upstream odoo {
        server 127.0.0.1:8016;
}
upstream odoochat {
        server 127.0.0.1:8078;
}

**Comment by gaikaz** [comment]
@susankc, could you do a simple test logged in to your server terminal:
`curl localhost:8016/xmlrpc/db`

It either gives you a html page for 404 error or some XML (the correct outcome)
If you get 404, then it really is some bad server setup and not the module.

**Comment by susankc** [comment]
i came up with these error 
<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 3.2 Final//EN">
<title>404 Not Found</title>
<h1>Not Found</h1>
<p>The requested URL was not found on the server.  If you entered the URL manually please check your spelling and try again.</p>
Did i do something wrong ?

**Comment by gaikaz** [comment]
No. It just means your setup is somehow wrong. I'm starting to believe you have the same problem like I did. You are not loading `base` module server-wide. How do you start your odoo server? With `systemd`? if so, could you send your `systemd` service file contents here? It should be somewhere in `/etc/systemd/system` if it's an Debian based system.

**Comment by susankc** [comment]
Yes you can see below
[Service]
Type=simple
PermissionsStartOnly=true
User=odoo
Group=odoo
SyslogIndentifier=odoo
ExecStart=/opt/odoo/odoo/venv/bin/python3 /opt/odoo12/odoo/odoo-bin --config=/etc/odoo.conf --logfile /var/log/odoo/odoo.log
WorkingDirectory=/opt/odoo/odoo/

[Install]
WantedBy=multi-user.target

**Comment by susankc** [comment]
Thank you guys i figured out my problem. I have save problem like @gaikaz  i put --load in my service file and now it is working fine in my server. This setup was not need in 11 version i guess.

**Comment by Yenthe666** [comment]
Awesome, good luck guys!

**Comment by gr22** [comment]
Hey @gaikaz this was exactly the issue in my case. base module was not being pre-loaded.  Appreciate your efforts.

**Comment by Yenthe666** [comment]
As it seems to have hit three people I've also added it in the important information section of the readme. Added in https://github.com/Yenthe666/auto_backup/commit/4f45b45bd72a7e66a312cfe0033df1ee65597947 - I hope that helps others to identify the problem faster.

**Comment by OpencloudPro** [comment]
> Good news everybody! We have a delivery to... I mean, I figured it out! 😃
> 
> The problem was incorrect use of `--load=...` launch argument.
> Because I use `letsencrypt` module, it has to be loaded server-wide, which is done with this argument.
> 
> My setup used: `--load=web,letsencrypt`
> 
> Module `web` here is specified, because it is a default of this argument and we still want to load it.
> On Odoo 10.0 the default was `web,web_kanban`.
> On Odoo 11.0 it was changed to `web`.
> On Odoo 12.0 it is now `base,web`
> 
> As we can see, Odoo 12.0 now has `base` loaded this way. It means, that module `base` used to be loaded server-wide (on all databases as a sort of a "singleton") automatically. Therefore database independent HTTP requests like `/xmlrpc/db` used to work no matter if database was selected.
> On my setup `base` was loaded on each database separately and that's why I wouldn't get a response from the server on `/xmlrpc/db` without first selecting a database. @gr22, tagging you here, as you probably have something similar done with your server 😉
> 
> **TL;DR:**
> If you are using `--load=...` launch argument with Odoo 12.0, you have to make sure you have `base` module loaded with it.

Thanks for the save!!
